backend/src/validation/mod.rs

```diff
@@ -0,0 +1,290 @@
+        // Validate URL format
+        if let Err(_) = Url::parse(&raw_event.url) {
+            return Err(ValidationError::InvalidUrl("Invalid URL format".to_string()));
+        }
+
+        // Validate IP address format
+        if IpAddr::from_str(ip_address).is_err() {
+            return Err(ValidationError::InvalidIpAddress("Invalid IP address format".to_string()));
+        }
+
+        // Validate timestamp is reasonable (config allows for some clock drift to account for packet latency)
+        if let Some(event_time) = DateTime::from_timestamp(raw_event.timestamp as i64, 0) {
+            let now = Utc::now();
+            let drift = (event_time - now).num_seconds().abs();
+            if drift > self.config.max_timestamp_drift_seconds {
+                return Err(ValidationError::InvalidTimestamp("Timestamp too far from current time".to_string()));
+            }
+        } else {
+            return Err(ValidationError::InvalidTimestamp("Invalid timestamp".to_string()));
+        }
+
+        Ok(())
+    }
+
+    /// Validate properties JSON structure
+    fn validate_properties_json(&self, properties: &str) -> Result<(), ValidationError> {
+        // For custom properties, we're more lenient with control characters
+        // (allow newlines/tabs in case someone tracks form texts)
+        // but we still disallow and reject dangerous control characters
+        if contains_dangerous_control_characters(properties) {
+            return Err(ValidationError::InvalidJson("Properties contain dangerous control characters".to_string()));
+        }
+
+        match serde_json::from_str::<serde_json::Value>(properties) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(ValidationError::InvalidJson(format!("Invalid JSON: {}", e))),
+        }
+    }
+
+    /// Log sanitized rejection details for debugging
+    fn log_sanitized_rejection(
+        &self,
+        error: &ValidationError,
+        raw_event: &RawTrackingEvent,
+        ip_address: &str,
+    ) {
+        // Hash IP address for privacy
+        let ip_hash = self.hash_ip_address(ip_address);
+        
+        // Sanitize URL to remove query parameters and fragments
+        let sanitized_url = self.sanitize_url(&raw_event.url);
+        
+        // Truncate user agent to prevent logging extremely long/malicious strings
+        let sanitized_user_agent = self.truncate_string(&raw_event.user_agent, 100);
+        
+        warn!(
+            rejection_reason = self.get_rejection_reason(error),
+            site_id = %raw_event.site_id,
+            event_name = %raw_event.event_name,
+            ip_hash = %ip_hash,
+            url_domain = %self.extract_domain(&raw_event.url),
+            url_path = %self.extract_path(&sanitized_url),
+            user_agent_prefix = %sanitized_user_agent,
+            properties_size = %raw_event.properties.len(),
+            "Event rejected by validation"
+        );
```

Comment:
This logs the actual control characters, which I don't think we want.

For instance, sending `window.betterlytics.event("hel\n\nlo\t\tworld")` will cause it to print:

<img width="1892" height="101" alt="image" src="https://github.com/user-attachments/assets/d7992177-54a5-4030-b0ef-22deefa290da" />